File: Programmers/Bridge.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

//이분 탐색
bool binary_search(int& n, int& k, vector<int>& v){
    int count = 0;
    
    for(int i = 0; i < v.size(); i++){
        if(v[i] - n <= 0)
            count++;
        else
            count = 0;
        if(count >= k)
            return true;
    }
    return false;
}

int solution(vector<int> stones, int k) {
    int first = 1, last = *max_element(stones.begin(), stones.end());
    
    while(first <= last){
        int mid = (first + last) / 2;
        if(binary_search(mid, k, stones))
            last = mid - 1;
        else
            first = mid + 1;
    }
    
    return first;
}

File: Programmers/Bridge.js
function check(stones,mid,k) {
    let n = 0;
    for(let i = 0; i < stones.length; i++) {
        if(stones[i] < mid) {
            n += 1;  
        }
        else
            n = 0;
        if(n >= k ) {
            return false;
        }
    }
    return true;
}

function solution(stones, k) {
    let left = 1;
    let right = 200000000;
    
    while(left < right-1) {
        let mid = parseInt((left + right) / 2);
        if(check(stones,mid,k)) {
            left = mid;
        }
        else
            right = mid;
    }
    
    return left;
}